sbncode/EventGenerator/MultiPart/MultiPartVertex_module.cc

```diff
@@ -197,182 +210,209 @@ MultiPartVertex::MultiPartVertex(fhicl::ParameterSet const & p)
-            num_part -= 1;
-        }
-        if(gen_count_v[idx] >= _param_v[idx].multi[1])
-        weight_v[idx] = 0.;
+  std::vector<size_t> result;
+  std::vector<size_t> gen_count_v(_param_v.size(),0);
+
+  int num_part = (int)(fFlatRandom->fire(_multi_min,_multi_max+1-1.e-10));
+
+  // generate min multiplicity first
+  std::vector<double> weight_v(_param_v.size(),0);
+  for(size_t idx=0; idx<_param_v.size(); ++idx) {
+    weight_v[idx] = _param_v[idx].weight;
+    for(size_t ctr=0; ctr<_param_v[idx].multi[0]; ++ctr) {
+      result.push_back(idx);
+      gen_count_v[idx] += 1;
+      num_part -= 1;
     }
+    if(gen_count_v[idx] >= _param_v[idx].multi[1])
+      weight_v[idx] = 0.;
+  }
 
-    assert(num_part >= 0);
+  assert(num_part >= 0);
 
-    while(num_part) {
+  while(num_part) {
 
-        double total_weight = 0;
-        for(auto const& v : weight_v) total_weight += v;
+    double total_weight = 0;
+    for(auto const& v : weight_v) total_weight += v;
 
-        double rval = 0;
-        rval = fFlatRandom->fire(0,total_weight);
+    double rval = 0;
+    rval = fFlatRandom->fire(0,total_weight);
 
-        size_t idx = 0;
-        for(idx = 0; idx < weight_v.size(); ++idx) {
-            rval -= weight_v[idx];
-            if(rval <=0.) break;
-        }
+    size_t idx = 0;
+    for(idx = 0; idx < weight_v.size(); ++idx) {
+      rval -= weight_v[idx];
+      if(rval <=0.) break;
+    }
 
-        // register to the output
-        result.push_back(idx);
+    // register to the output
+    result.push_back(idx);
 
-        // if generation count exceeds max, set probability weight to be 0
-        gen_count_v[idx] += 1;
-        if(gen_count_v[idx] >= _param_v[idx].multi[1])
-            weight_v[idx] = 0.;
+    // if generation count exceeds max, set probability weight to be 0
+    gen_count_v[idx] += 1;
+    if(gen_count_v[idx] >= _param_v[idx].multi[1])
+      weight_v[idx] = 0.;
 
-        --num_part;
-    }
-    return result;
+    --num_part;
+  }
+  return result;
 }
 
 
 void MultiPartVertex::GenPosition(double& x, double& y, double& z) {
 
-    x = fFlatRandom->fire(_xrange[0],_xrange[1]);
-    y = fFlatRandom->fire(_yrange[0],_yrange[1]);
-    z = fFlatRandom->fire(_zrange[0],_zrange[1]);
+  x = fFlatRandom->fire(_xrange[0],_xrange[1]);
+  y = fFlatRandom->fire(_yrange[0],_yrange[1]);
+  z = fFlatRandom->fire(_zrange[0],_zrange[1]);
 
-    if(_debug>0) {
-        std::cout << "Generating a rain particle at ("
-            << x << "," << y << "," << z << ")" << std::endl;
-    }
+  if(_debug>0) {
+    std::cout << "Generating a rain particle at ("
+	      << x << "," << y << "," << z << ")" << std::endl;
+  }
+}
+
+void MultiPartVertex::GenBoost(double& bx, double& by, double& bz) {
+
+  double gbmag = 0.;
+  if ( _gamma_beta_range[1] > _gamma_beta_range[0]) {
+    gbmag = fFlatRandom->fire(_gamma_beta_range[0],_gamma_beta_range[1]);
+  }
+  else {
+    gbmag = _gamma_beta_range[0];
+  }
```

Comment:
Should these checks be done when first reading the cfg? Also for the else, do we just want to select the lowest one or should we assert the second element is larger than the first?

Reply:
i see that other places default to the first element so i bet this is fine.